File: LeoHuang/238-Production-of-Array-Except-Self/238-Production-of-Array-Except-Self.py
class Solution(object):
  def productExceptSelf(self, nums):
    """
    :type nums: List[int]
    :rtype: List[int]
    """
    # calculate the product of all items in nums
    totalProduct = reduce((lambda total, num: total * num), nums)
    if totalProduct:
      return map((lambda num: totalProduct / num), nums)
    else:
      totalProduct = []
      for i in range(len(nums)):
        totalProduct.append(reduce((lambda total, num: total * num), \
						[num for index, num in enumerate(nums) if index != i]))
      return totalProduct

File: LeoHuang/238-Production-of-Array-Except-Self/238-Production-of-Array-Except-Self.js
/**
 * @param {number[]} nums
 * @return {number[]}
 */
var productExceptSelf = function(nums) {
  // make the production of all items in nums
  let totalProduct = nums.reduce(function(total, num) {
    return total * num;
  });
  // check whether nums contains 0 or not
  if(totalProduct) {
    return nums.map(num => {
      return totalProduction / num;
    });
  } else {
    // if nums contains 0 
    // use methods not in O(n)
    totalProduct = [];
    nums.forEach(function(currentValue, index, arr) {
      totalProduction.push(arr.filter(function(item, idx) {
        return idx != index;
      }).reduce(function(total, num) {
        return total * num;
      }));
    });
    return totalProduct;
  }
};
